File: factorial/tail_recursive_factorial.rb
def factorial(total, n)
  if n == 0
    return total
  end

  factorial(total * n, n - 1)
end

puts factorial(1, 0)
puts factorial(1, 1)
puts factorial(1, 2)
puts factorial(1, 3)
puts factorial(1, 4)
puts factorial(1, 5)

File: challenges/factorial/tail_recursive_factorial.go
package main

import "fmt"

func factorial(total uint, n uint) uint {
	if n == 0 {
		return total
	}

	return factorial(total*n, n-1)
}

func main() {
	fmt.Println(factorial(1, 0))
	fmt.Println(factorial(1, 1))
	fmt.Println(factorial(1, 2))
	fmt.Println(factorial(1, 3))
	fmt.Println(factorial(1, 4))
	fmt.Println(factorial(1, 5))
}
